contracts/DXswapFeeReceiver.sol

```diff
@@ -6,122 +6,226 @@ import './interfaces/IWETH.sol';
-        uint denominator = reserveIn.mul(10000).add(amountInWithFee);
-        uint amountOut = numerator / denominator;
-        
-        TransferHelper.safeTransfer(
-            fromToken, address(pairToUse), amountIn
-        );
-        
-        (uint amount0Out, uint amount1Out) = fromToken < WETH ? (uint(0), amountOut) : (amountOut, uint(0));
-        
-        pairToUse.swap(
-            amount0Out, amount1Out, address(this), new bytes(0)
-        );
-        
-        IWETH(WETH).withdraw(amountOut);
-        TransferHelper.safeTransferETH(ethReceiver, amountOut);
+
+        (uint256 reserve0, uint256 reserve1, ) = pairToUse.getReserves();
+        (uint256 reserveIn, uint256 reserveOut) = fromToken < WETH ? (reserve0, reserve1) : (reserve1, reserve0);
+
+        require(reserveIn > 0 && reserveOut > 0, 'DXswapFeeReceiver: INSUFFICIENT_LIQUIDITY'); // should never happen since pool was checked before
+        uint256 amountInWithFee = amountIn.mul(uint256(10000).sub(pairToUse.swapFee()));
+        uint256 numerator = amountInWithFee.mul(reserveOut);
+        uint256 denominator = reserveIn.mul(10000).add(amountInWithFee);
+        amountOut = numerator / denominator;
+
+        TransferHelper.safeTransfer(fromToken, address(pairToUse), amountIn);
+
+        (uint256 amount0Out, uint256 amount1Out) = fromToken < WETH ? (uint256(0), amountOut) : (amountOut, uint256(0));
+
+        pairToUse.swap(amount0Out, amount1Out, address(this), new bytes(0));
     }
 
-    // Transfer to the owner address the token converted into ETH if possible, if not just transfer the token.
-    function _takeETHorToken(address token, uint amount) internal {
-      if (token == WETH) {
-        // If it is WETH, transfer directly to ETH receiver
-        IWETH(WETH).withdraw(amount);
-        TransferHelper.safeTransferETH(ethReceiver, amount);
-      } else if (isContract(pairFor(token, WETH))) {
-        // If it is not WETH and there is a direct path to WETH, swap and transfer WETH to ETH receiver
-        _swapTokensForETH(amount, token);
-      } else {
-        // If it is not WETH and there is not a direct path to WETH, transfer tokens directly to fallback receiver
-        TransferHelper.safeTransfer(token, fallbackReceiver, amount);
-      }
+    // Helper function to know if token-WETH pool exists and has enough liquidity
+    function _isSwapPossible(
+        address token0,
+        address token1,
+        uint256 amount
+    ) internal view returns (bool) {
+        address pair = pairFor(token0, token1);
+        if (!isContract(pair)) return false;
+
+        (uint256 reserve0, uint256 reserve1, ) = IDXswapPair(pair).getReserves();
+        (uint256 reserveIn, uint256 reserveOut) = token0 < token1 ? (reserve0, reserve1) : (reserve1, reserve0);
+        if (reserveIn == 0 || reserveOut == 0) return false;
+
+        uint256 priceImpact = amount.mul(10000) / reserveIn; // simplified formula
```

Comment:
I see this is intentionally a simplified formula.  Do we think it's good enough?

Reply:
I've done quite a lot of calculations (if you're interested in details I can share the sheet) and imo it's good enough just as an extra protection step. Ofc it is possible to implement more complex formula, but that could cost more. I'll check it out and compare gas usage.

Reply:
I'd like to see more if possible.